### PR TITLE
feat(base_types): Add `Wei` type

### PR DESCRIPTION
**Pull request by marioevz**
## 🗒️ Description
Adds a type called `Wei` which allows string parsing different combinations of fund amount strings, such as `"1 ether"`, `"100 gwei"`, `"10**9 wei"`, and others.

Mostly meant to be used in command parameter parsing to allow more human-readable commands.

## 🔗 Related Issues
None

## ✅ Checklist
- [x] All: Set appropriate labels for the changes.
- [x] All: Considered squashing commits to improve commit history.
- [x] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [x] Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been added to [converted-ethereum-tests.txt](/ethereum/execution-spec-tests/blob/main/converted-ethereum-tests.txt).
- [x] Tests: A PR with removal of converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been opened.
- [x] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

**Review by spencer-tb — APPROVED** [review]
Nice! Never knew "finney" or "szabo" existed!

File: src/ethereum_test_base_types/base_types.py [hunk 1]
```diff
@@ -75,6 +75,50 @@ def or_none(cls: Type[N], input: N | NumberConvertible | None) -> N | None:
...
+                value = int(float(value_str))
+            elif "**" in value_str:
+                base, exp = value_str.split("**")
+                value = int(base) ** int(exp)
+            else:
+                value = int(value_str)
+            return super(Number, cls).__new__(cls, value * multiplier)
+        return super(Number, cls).__new__(cls, to_number(input))
+
+
```
Comment (danceratopz) [inline src/ethereum_test_base_types/base_types.py:121]: Allow synonyms:
```suggestion
    @staticmethod
    def _get_multiplier(unit: str) -> int:
        """
        Returns the multiplier for the given unit of wei, handling synonyms.
        """
        match unit:
            case "wei":
                return 1
            case "kwei" | "babbage" | "femtoether":
                return 10**3
            case "mwei" | "lovelace" | "picoether":
                return 10**6
            case "gwei" | "shannon" | "nanoether" | "nano":
                return 10**9
            case "szabo" | "microether" | "micro":
                return 10**12
            case "finney" | "milliether" | "milli":
                return 10**15
            case "ether":
                return 10**18
            case _:
                raise ValueError(f"Invalid unit {unit}")
```

File: src/ethereum_test_base_types/base_types.py [hunk 2]
```diff
@@ -75,6 +75,50 @@ def or_none(cls: Type[N], input: N | NumberConvertible | None) -> N | None:
...
+                elif unit == "gwei":
+                    multiplier = 10**9
+                elif unit == "szabo":
+                    multiplier = 10**12
+                elif unit == "finney":
+                    multiplier = 10**15
+                elif unit == "ether":
+                    multiplier = 10**18
+                else:
+                    raise ValueError(f"Invalid unit {unit}")
```
Comment (danceratopz) [inline src/ethereum_test_base_types/base_types.py:109]: Normalize with `lower()`; allow upper/mixed case strings:
```suggestion
            if len(words) > 1:
                unit = words[1].lower()
                multiplier = cls._get_multiplier(unit)
```

File: src/ethereum_test_base_types/tests/test_base_types.py [hunk 3]
```diff
@@ -52,3 +52,25 @@ def test_comparisons(a: Any, b: Any, equal: bool):
...
         assert not a == b
+
+
+@pytest.mark.parametrize(
+    "s, expected",
+    [
+        ("0", 0),
+        ("10**18", 10**18),
+        ("1e18", 10**18),
+        ("1 ether", 10**18),
```
Comment (danceratopz) [inline src/ethereum_test_base_types/tests/test_base_types.py:63]:
```suggestion
        ("1 ether", 10**18),
        ("2 ether", 2*10**18),
        ("2.1 ether", 2.1*10**18),
        ("2.1 Ether", 2*10**18),
        ("2.1 ETHER", 2*10**18),
```

File: src/ethereum_test_base_types/tests/test_base_types.py [hunk 4]
```diff
@@ -52,3 +52,25 @@ def test_comparisons(a: Any, b: Any, equal: bool):
...
+        ("1 wei", 1),
+        ("1 gwei", 10**9),
+        ("1 szabo", 10**12),
+        ("1 finney", 10**15),
+        ("1 kwei", 10**3),
+        ("1 mwei", 10**6),
+    ],
```
Comment (danceratopz) [inline src/ethereum_test_base_types/tests/test_base_types.py:70]:
```suggestion
        ("1 babbage", 10**3),
        ("1 femtoether", 10**3),
        ("1 Lovelace", 10**6),
        ("1 Picoether", 10**6),
        ("1 gwei", 10**9),
        ("1 shannon", 10**9),
        ("1 nanoether", 10**9),
        ("1 nano", 10**9),
        ("1 microether", 10**12),
        ("1 micro", 10**12),
        ("1 milliether", 10**15),
        ("1 milli", 10**15),
    ],
```

**Review by danceratopz — APPROVED** [review]
Really nice, few suggestions below!

Perhaps you'd prefer to make it explicit and not allow synonyms, I don't have a strong opinion :slightly_smiling_face:  but normalizing the input unit and using `match` in a helper `_get_multiplier()` method makes sense to me.

File: src/ethereum_test_base_types/tests/test_base_types.py [hunk 5]
```diff
@@ -52,3 +52,25 @@ def test_comparisons(a: Any, b: Any, equal: bool):
...
+
+
+@pytest.mark.parametrize(
+    "s, expected",
+    [
+        ("0", 0),
+        ("10**18", 10**18),
+        ("1e18", 10**18),
+        ("1 ether", 10**18),
+        ("1 wei", 1),
```
Comment (danceratopz) [inline src/ethereum_test_base_types/tests/test_base_types.py:64]:
```suggestion
        ("10**9 wei", 10**9),
```

**Review by danceratopz — COMMENTED** [review]
Another test case.

**Pull request merged**